docs/standard/native-interop/tutorial-custom-marshaller.md

```diff
@@ -0,0 +1,278 @@
+        }
+    }
+}
+```
+
+The conversion from `ErrorDataUnmanaged` to `ErrorData` is the inverse of what you did for the "in" mode in the preceding code. Remember that you also need to "clean up" any allocations that the unmanaged environment expected you to perform.
+
+For the managed to unmanaged "out" marshaller, you're going to do something special. The name of the data type you're marshalling is called `error_data` and .NET typically expresses errors as exceptions. However, some errors are more impactful than others. Errors identified as "fatal" usually indicate a catastrophic or unrecoverable error. Notice the `error_data` has a field to check if the error is fatal. You'll marshal an `error_data` into managed code, and if it's fatal, you'll throw an exception rather than just converting it into an `ErrorData` and returning it.
+
+```csharp
+namespace CustomMarshalling
+{
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedIn, typeof(ErrorDataMarshaller))]
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ElementOut, typeof(Out))]
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedOut, typeof(ThrowOnFatalErrorOut))]
+    internal static unsafe class ErrorDataMarshaller
+    {
+        //
+        // Other marshallers removed
+        //
+
+        public static class ThrowOnFatalErrorOut
+        {
+            public static ErrorData ConvertToManaged(ErrorDataUnmanaged unmanaged)
+                => throw new NotImplementedException();
+
+            public static void Free(ErrorDataUnmanaged unmanaged)
+                => throw new NotImplementedException();
+        }
+    }
+}
+```
+
+An "out" parameter converts from an unmanaged context into a managed context, so you implement the `ConvertToManaged` method. When the unmanaged callee returns and provides an `ErrorDataUnmanaged` object, you can inspect it using your `ElementOut` mode marshaller and then check if it's marked as a "fatal" error. If so, that's your indication to throw as opposed to just returning the `ErrorData`.
+
+```csharp
+public static ErrorData ConvertToManaged(ErrorDataUnmanaged unmanaged)
+{
+    ErrorData data = Out.ConvertToManaged(unmanaged);
+    if (data.IsFatalError)
+        throw new ExternalException(data.Message, data.Code);
+
+    return data;
+}
+```
+
+Perhaps you're not only going to consume the native library, but you also want to share your work with the community and provide an interop library. You can provide `ErrorData` with an implied marshaller whenever it's used in a P/Invoke by adding `[NativeMarshalling(typeof(ErrorDataMarshaller))]` to the `ErrorData` definition. Now, anyone using your definition of this type in a `LibraryImport` call will get the benefit of your marshallers. They can always override your marshallers by using `MarshalUsing` at the use site.
+
+```csharp
+[NativeMarshalling(typeof(ErrorDataMarshaller))]
+struct ErrorData { ... }
+```
+
+  <!-- links -->
```

Comment:
Do we want a `See also` section with these links (and probably pinvoke-source-generation.md and custom-marshalling-source-generation.md)?

Reply:
These links are embedded throughout the document. I don't think they need to be called out specifically. I will cross reference `pinvoke-source-generation.md`, which seems appropriate.